Feature/config manager and routes

DarkKlaine opened the pull request.

DarkKlaine committed:
DKE-8 Обновил WebTemplateEngine.php под новые требования https://u.to/DbsgIA

DarkKlaine committed:
DKE-8 Добавил файловую структуру для модуля UsersCRUD https://u.to/DbsgIA

DarkKlaine committed:
DKE-8 Промежуточный комит

DarkKlaine committed:
DKE-8 Промежуточный комит

DarkKlaine committed:
DKE-8 Эксперименты с конфиг менеджером. Помежуточный коммит. https://u.to/DbsgIA

DarkKlaine committed:
DKE-8 Эксперименты с конфиг менеджером. Помежуточный коммит. https://u.to/DbsgIA

DarkKlaine committed:
DKE-8 Допилил конфиг менеджер, вроде

DarkKlaine committed:
DKE-8 Допилил конфиг менеджер, вроде

DarkKlaine committed:
DKE-8 Допилил конфиг менеджер, вроде

DarkKlaine committed:
DKE-8 Добавил главную страницу движка

DarkKlaine committed:
DKE-8 Исправил косяки

DarkKlaine committed:
DKE-8 Представил класс и метод в виде callable в роутах.

DarkKlaine committed:
DKE-8 Поправил WebRouter.php.

brainraider commented on Config/WebCfg/app.php:
```diff
@@ -3,7 +3,7 @@
     'homeUrl' => '/',
     'authEnabled' => true,
-    'authSessionLifeTime' => 300,
+    'authSessionLifeTime' => 3600,
```
Я бы рекомендовал сделать как делают взрослые дяди )
Делаешь жизнь сессии, например, 15 минут.
И каждый раз при обращении обновляешь время жизни сессии.

Таким образом получается, что сессия умрет только через 15 минут АФК

DarkKlaine replied:
ТЗ было чтоб сессия жила 5 минут. Переделаю.

DarkKlaine replied:
Сделал https://github.com/DarkKlaine/Php_Calculator/pull/4/commits/a68cb4a1bd199e7d5e31b7abc7b8e9fd18a31310

brainraider commented on Config/WebCfg/authWhitelist.php:
```diff
@@ -1,6 +1,7 @@
 
 return [
+    'engineHome' => '/',
     'accessDenied' => '/accessDenied',
```
Кстати, обычно делают общую страницу для всех видов ошибок (404, 403 и т.п.).
В контроллерах этих страниц ловят исключение и исходя из него решают, какую вьюху отрисовывать

DarkKlaine replied:
Буду делать отдельную ветку для исключений. Отложил пока.

DarkKlaine replied:
https://darkklaine.youtrack.cloud/issue/DKE-4

brainraider commented on Config/WebCfg/engineRoutes.php:
```diff
@@ -0,0 +1,19 @@
+return [
+    'EngineHome' => [
+        'url' => '/',
+        'action' => [IEngineControllerWeb::class, 'engineHomePage']
```
:D

brainraider commented on Config/containerConfig.php:
```diff
@@ -0,0 +1,20 @@
+$config = require __DIR__ . '/modules.php';
+$modulesPath = $config['modulesPath'];
+$modulesList = $config['modulesList'];
+$moduleDepsPath = '/Config/container.di.php';
```
Ну, Deps так Deps )

DarkKlaine replied:
Упс, забыл, ты был прав, надо сразу писать правильно

DarkKlaine replied:
Исправил https://github.com/DarkKlaine/Php_Calculator/pull/4/commits/b7643c91364631c382d17c0faf0c4aa72e9b9d9f

brainraider commented on Config/modules.php:
```diff
@@ -0,0 +1,68 @@
+];
+
+
+/** BaseCfgManager
```
Может этому не место в коде проекта? ) Хватит носки грязные по квартире раскидывать )

DarkKlaine replied:
Исправил https://github.com/DarkKlaine/Php_Calculator/pull/4/commits/a78b580e3556b1763dd46ddf89a5185afb8ab41c

brainraider commented on Engine/Services/ConfigManagers/AuthConfigManagerWeb.php:
```diff
@@ -2,37 +2,27 @@
         $this->authEnabled = $appConfig['authEnabled'];
         $this->authSessionLifeTime = $appConfig['authSessionLifeTime'];
         $this->accessDeniedPage = $appConfig['authWhitelist']['accessDenied'];
         $this->authWhitelist = $appConfig['authWhitelist'];
```
ИМХО, это информация не в AUTH должна быть, а где-то в контексте роутинга.
Типа, страницы, которые не требуют авторизации.

DarkKlaine replied:
Ок, подумаю над этим когда буду перепиливать Auth под роли пользователей.

brainraider commented on Engine/Services/Routers/WebRouter/WebRouter.php:
```diff
@@ -2,62 +2,65 @@
-            $route['action'] ?? '',
-            $url,
-        );
+        foreach ($routes as $key => $value) {
```
Представь, что у тебя 1000 зареганных путей )
Каждый раз перебирать 1000 элементов - плохой вариант.

DarkKlaine replied:
Представил, ужаснулся. Но не сильно. Подумаю на досуге как можно и удобно и без цикла это сделать.

brainraider commented on Engine/Services/Routers/WebRouter/WebRouter.php:
```diff
@@ -2,62 +2,65 @@
+            $controller = $this->container->get($className);
+            $controller->$methodName($request);
+        } else {
+            $message = "Ошибка в WebRouter. Неправильный 'action' в *Routes.php.";
```
Исключения, Дмитрий! Пора использовать исключения!

DarkKlaine replied:
Буду делать отдельную ветку для исключений. Отложил пока.

DarkKlaine replied:
https://darkklaine.youtrack.cloud/issue/DKE-4

brainraider commented on Engine/Views/WebTemplateEngine.php:
```diff
@@ -7,16 +7,20 @@ class WebTemplateEngine implements IWebTemplateEngine
     }
 
-    public function setInjectTplFile(string $injectFile): void
+    public function setInjectTplFile(?string $linksTpl, string $contentTpl, ?string $scriptTpl): void
```
Лучше размещать необязательные параметры в конце.

Т.е. твой метод может выглядеть так:

```
public function setInjectTplFile(string $contentTpl, ?string $linksTpl = null, ?string $scriptTpl = null): void
```

И там, где тебе не нужно вставлять линки и скрипты ты просто можешь делать так:

```
$engine->setInjectTplFile($contentTpl);
```

Ну и название у метода какое-то стремное :)

DarkKlaine replied:
Исправил метод https://github.com/DarkKlaine/Php_Calculator/pull/4/commits/36d514ddc682a5ecf75ba10e8738a96828bfe340

> Ну и название у метода какое-то стремное :)

Согласен, подумаю на досуге что можно с этим сделать.

brainraider commented on Modules/Calculator/Config/webRoutes.php:
```diff
@@ -0,0 +1,27 @@
+use Modules\Calculator\Controllers\WebControllers\IWebHistoryController;
+
+return [
+    'Calculator' => [
```
Вот тут не хватает еще разрешенных методов обращения.
У тебя сейчас работает и по POST и по GET и даже по DELETE методу.
Нужно, чтобы каждый метод работал только по нужным, определенным методам.

Выдать инфу - GET
Записать инфу (или посчитать) - POST
Изменить - PUT
Удалить - DELETE

И т.д. 
Обычно так делают

DarkKlaine replied:
Ничо не понял, но создал задачу https://darkklaine.youtrack.cloud/issue/DKE-10

brainraider commented on Modules/Calculator/Controllers/WebControllers/WebCalculatorController.php:
```diff
@@ -56,6 +56,7 @@ public function calculate(WebRequestDTO $request): void
         $encodedInput = urlencode($inputDataString);
         $encodedResult = urlencode($result);
-        $this->redirectHandler->redirect(sprintf("/?input=%s&result=%s", $encodedInput, $encodedResult));
+        $url = $this->calculatorUrl . sprintf("/?input=%s&result=%s", $encodedInput, $encodedResult);
```
Это выглядит как костыль.
Есть специальные функции для сборки параметров запроса (GET) и для сборки тела запроса тоже (POST).

Ну и напрашивается какой-то сервис для работы с URL.

DarkKlaine replied:
Убрал костыль https://github.com/DarkKlaine/Php_Calculator/pull/4/commits/ada952a7c4e39207cc6913adb58af8f1fb8d16a9
Задача по сервису https://darkklaine.youtrack.cloud/issue/DKE-9

brainraider commented on Modules/Calculator/Views/Templates/links.tpl.php:
```diff
@@ -0,0 +1,14 @@
+<button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent1" aria-controls="navbarSupportedContent1" aria-expanded="false" aria-label="Toggle navigation"><span class="navbar-toggler-icon"></span>
```
Такие файлы обычно называют не "ссылки", а "меню" )
Изначально я думал, что речь идет про linc для css )

DarkKlaine replied:
Переименовал https://github.com/DarkKlaine/Php_Calculator/pull/4/commits/36d514ddc682a5ecf75ba10e8738a96828bfe340

brainraider reviewed (commented):
В целом всё круто! 
Есть конечно, куда стремиться, но уже сейчас выглядит хорошо!

DarkKlaine committed:
Исправил https://github.com/DarkKlaine/Php_Calculator/pull/4#discussion_r1409512137

DarkKlaine committed:
Исправил https://github.com/DarkKlaine/Php_Calculator/pull/4#discussion_r1409516003

DarkKlaine committed:
Исправил https://github.com/DarkKlaine/Php_Calculator/pull/4#discussion_r1409516890

DarkKlaine committed:
Убрал костыль :) https://github.com/DarkKlaine/Php_Calculator/pull/4#discussion_r1409553544

DarkKlaine committed:
Исправил https://github.com/DarkKlaine/Php_Calculator/pull/4#discussion_r1409543565
Переименовал https://github.com/DarkKlaine/Php_Calculator/pull/4#discussion_r1409555186
